File: lessons/ads/native_cache.py
from typing import Any, Optional


class NativeCache:
    def __init__(self, sz):
        self.size = sz
        self.slots = [None] * self.size
        self.values = [None] * self.size
        self.hits = [0] * self.size

    def hash_fun(self, key: str) -> int:
        # в качестве key поступают строки!
        # всегда возвращает корректный индекс слота
        return (
            sum([v * (i + 1) for i, v in enumerate(bytearray(key, encoding="utf-8"))])
            % self.size
        )

    def _seek_slot(self, key: str) -> Optional[int]:
        # находит индекс пустого слота для значения, или None
        slot = self.hash_fun(key)
        first_slot = slot
        minor_slot = slot
        while self.slots[slot] is not None and self.slots[slot] != key:
            minor_slot = slot if self.hits[minor_slot] > self.hits[slot] else minor_slot
            slot = (slot + 1) % self.size
            if slot == first_slot:
                return minor_slot
        return slot

    def _find(self, key: str) -> Optional[int]:
        # находит индекс слота со значением, или None
        slot = self.hash_fun(key)
        first_slot = slot
        while self.slots[slot] != key:
            slot = (slot + 1) % self.size
            if slot == first_slot:
                return None
        return slot

    def put(self, key: str, value: Any) -> None:
        # гарантированно записываем
        # значение value по ключу key
        index = self._seek_slot(key)
        self.slots[index] = key
        self.values[index] = value
        self.hits[index] += 1

    def get(self, key: str) -> Optional[Any]:
        # возвращает value для key,
        # или None если ключ не найден
        index = self._find(key)
        if index is None or self.slots[index] != key:
            return None
        return self.values[index]

File: lessons/ads3_go/NativeCache/NativeCache.go
// 12
package main

import (
	"errors"
)

type NativeCache[T any] struct {
	size   int
	slots  []string
	values []T
	hits   []int
	step   int
}

// создание экземпляра словаря
func Init[T any](sz int) NativeCache[T] {
	nc := NativeCache[T]{size: sz, slots: nil, values: nil, hits: nil, step: 1}
	nc.slots = make([]string, sz)
	nc.values = make([]T, sz)
	nc.hits = make([]int, sz)
	return nc
}

func (nc *NativeCache[T]) HashFun(value string) int {
	// всегда возвращает корректный индекс слота
	// сумма байтов по модулю размера таблицы - из предыдущего занятия
	sum := 0
	for _, b := range []byte(value) {
		sum += int(b)
	}
	return sum % nc.size
}

func (nc *NativeCache[T]) seekMinHitSlot() int {
	minInd := 0
	min := nc.hits[minInd]

	for i, v := range nc.hits {
		if v < min {
			min = v
			minInd = i
		}
	}
	nc.hits[minInd] = 0
	return minInd
}

func (nc *NativeCache[T]) seekSlot(key string) int {
	// находит индекс пустого слота для значения,
	// или -1 если свободного слота не найти
	hash := nc.HashFun(key)

	for i := 0; i < nc.size; i++ {
		index := (hash + i*nc.step) % nc.size
		if nc.slots[index] == "" {
			return index
		}
	}

	// если не нашли ни одного пустого слота - надо освободить слот и вернуть его индекс
	return nc.seekMinHitSlot()
}

func (nc *NativeCache[T]) find(key string) int {
	// находит индекс слота со значением, или -1
	hash := nc.HashFun(key)
	for i := 0; i < nc.size; i++ {
		index := (hash + i*nc.step) % nc.size
		if nc.slots[index] == key {
			return index
		}
	}
	return -1
}

func (nc *NativeCache[T]) IsKey(key string) bool {
	// возвращает true если ключ имеется
	slot := nc.find(key)
	return slot != -1
}

func (nc *NativeCache[T]) Get(key string) (T, error) {
	// возвращает value для key,
	// или error если ключ не найден

	var result T

	slot := nc.find(key)
	if slot == -1 {
		return result, errors.New("Key error!")
	}
	nc.hits[slot]++
	return nc.values[slot], nil
}

func (nc *NativeCache[T]) Put(key string, value T) {
	// гарантированно записываем
	// значение value по ключу key

	existingSlot := nc.find(key)
	if existingSlot != -1 {
		// считаем, что мы заинтересованы в частоте обращения к ключу, а не к значению
		// поэтому счетчик не трогаем
		nc.values[existingSlot] = value
		return
	}

	slot := nc.seekSlot(key)

	if slot == -1 {
		return
	}
	nc.slots[slot] = key
	nc.values[slot] = value
}
